sdk/src/signing.rs

```diff
@@ -208,42 +228,96 @@ pub async fn tx_signers<'a>(
 /// hashes needed for monitoring the tx on chain.
 ///
 /// If it is a dry run, it is not put in a wrapper, but returned as is.
-pub fn sign_tx<U: WalletIo>(
-    wallet: &mut Wallet<U>,
+pub async fn sign_tx<'a, F: std::future::Future<Output = Result<Tx, Error>>>(
+    context: &impl Namada<'a>,
     args: &args::Tx,
     tx: &mut Tx,
     signing_data: SigningTxData,
+    sign: impl Fn(Tx, common::PublicKey, HashSet<Signable>) -> F,
 ) -> Result<(), Error> {
+    let mut used_pubkeys = HashSet::new();
+
+    // First try to sign the raw header with the supplied signatures
     if !args.signatures.is_empty() {
         let signatures = args
             .signatures
             .iter()
-            .map(|bytes| SignatureIndex::deserialize(bytes).unwrap())
+            .map(|bytes| {
+                let sigidx = SignatureIndex::deserialize(bytes).unwrap();
+                used_pubkeys.insert(sigidx.pubkey.clone());
+                sigidx
+            })
             .collect();
         tx.add_signatures(signatures);
-    } else if let Some(account_public_keys_map) =
-        signing_data.account_public_keys_map
+    }
+
+    // Then try to sign the raw header with private keys in the software wallet
+    if let Some(account_public_keys_map) = signing_data.account_public_keys_map
```

Comment:
Are we possibly signing with multiple keys? I'm a little bit confused by this flow. It seems to me that we should first lookup the key pair that the user instructed us to sign with, then use it (or return an error if not found)

Reply:
Agreed, this flow is a little confusing. Yes, we are potentially signing with multiple keys. I guess either way is possible: first try to apply the signatures the user has supplied via the CLI (used in multisig contexts), and then try signing with the keys the user has specified. Or the other way round. If wanted to immediately error on a key pair not being found, we would have to do some refactoring in order to keep the multisig implementation/flow working.

Reply:
Ok, gotcha. Let's think about this later/separately.